### ISSUE TITLE
Mark ExceedDeadline Job with completions == succeed instead of  Failed?

### ISSUE THREAD

**Issue by weiwei04** [kubernetes/kubernetes#46497]
**this a FEATURE REQUEST**:

Should we mark exceedDeadline Job(completions: 1/parallelism: 1) with 1 Pod succeed as Completed instead of Failed? I think it's a little wired, my expected Pod(completions: 1/parallelism: 1) has succeed, but the Job is failed.

**Kubernetes version** (use `kubectl version`): version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1"

**Environment**:
- **Cloud provider or hardware configuration**: baremetal
- **OS** (e.g. from /etc/os-release): 
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
- **Kernel** (e.g. `uname -a`):
Linux xs342 4.4.0-72-generic #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**What happened**:
I use the job.yaml as below

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: test-job-deadline
  labels:
    name: test-job-deadline
spec:
  activeDeadlineSeconds: 5
  completions: 1
  template:
    metadata:
      labels:
        name: test-job-deadline
    spec:
      containers:
      - name: test
        image: busybox:1.26.2-glibc
        args:
        - sleep 10 && echo done && exit 0
        command:
        - sh
        - -c
        imagePullPolicy: IfNotPresent
      restartPolicy: Never
```

and got job Failed with 1 Pod succeeded, here's the kubectl get job
```yaml
status:
  conditions:
  - lastProbeTime: 2017-05-26T06:19:46Z
    lastTransitionTime: 2017-05-26T06:19:46Z
    message: Job was active longer than specified deadline
    reason: DeadlineExceeded
    status: "True"
    type: Failed
  startTime: 2017-05-26T06:19:33Z
  succeeded: 1
```

**What you expected to happen**:

Should we consider Job(completions: 1/parallelism: 1) with 1 Pod succeed as Completed even the Job has exceed the activeDeadline?

I spend some time to read the code, It appears because:
the Pod success kick a Update Event, and the JobController syncJob("test-job-deadline") first check the job [pastActiveDeadline](https://github.com/kubernetes/kubernetes/blob/release-1.6/pkg/controller/job/jobcontroller.go#L334) and mark the Job Failed.

If i set 'sleep SOMETIME && echo done && exit 0' the SOMETIME is longer than 2 * min-resync-period, the result was as expected Job Failed with succeed: 0, failed: 1, that's because the Event was kicked by informers Sync.

**How to reproduce it** (as minimally and precisely as possible):

kubectl apply -f job.yaml

**Comment by weiwei04** [comment]
Changed the title from "Mark ExceedDeadline Job(completions: 1/parallelism: 1) with 1 Pod succeed as Completed instead of  Failed?" to "Mark ExceedDeadline Job with completions == succeed instead of  Failed?".

Consider a scenario, the jobSync was kick by the last succeed Pod in a Job, jobSync checked the job has past the actibeDeadlineSeconds and mark the Job condition as Failed, but the job .spec.completions == succeed pod count. Will it be better if we mark the Job as succeed instead of Failed?

**Comment by soltysh** [comment]
Your fix provided in #46499 will only partially solve your use case. Unfortunately, we still need to solve the problem of short ADS, see https://github.com/kubernetes/kubernetes/issues/32149 for more details. If you're interested in fixing the other one, I'd be more than happy to review your PR :)

**Comment by weiwei04** [comment]
@soltysh  Thanks for the reply, I'd like to take a try to fix #32149.

**Comment by weiwei04** [comment]
fixed by #32149, close issue.
